packages/coding-standards/src/Sniffs/ForceLateStaticBindingForProtectedConstantsSniff.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\CodingStandards\Sniffs;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use SlevomatCodingStandard\Helpers\AnnotationHelper;
+use SlevomatCodingStandard\Helpers\ConstantHelper;
+use SlevomatCodingStandard\Helpers\TokenHelper;
+
+class ForceLateStaticBindingForProtectedConstantsSniff implements Sniff
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function register(): array
+    {
+        return [\T_CLASS];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $file, $classPosition)
+    {
+        $tokens = $file->getTokens();
+
+        $protectedConstants = $this->getAllProtectedConstantsInClass($file);
+
+        $selfPositions = TokenHelper::findNextAll($file, \T_SELF, $classPosition);
+
+        foreach ($selfPositions as $selfPosition) {
+            $doubleColonPosition = TokenHelper::findNextEffective($file, $selfPosition + 1);
```

Comment:
I think about extracting the long chain of IFs to function `isSelfAndShouldBeStatic`

and later in code `$constantName = $tokens[$stringPosition]['content'];` change to function `$constantName = $this->findConstatntName(...)`. It would maybe contain duplicit logic with `isSelf`, but this duplicit could be maybe also extracted

By this change you'll remove temporary variables `$doubleColonPosition, $stringPosition, $positionAfterString`

Reply:
I rewrote it, but not exactly as you suggested, because I didn't want to do the same job twice (going through constant candidates, check and get the position of constant name).

Hope it's ok. 

Found a bug in the proccess, by the way, so I added another test case to cover the situations, where `self` is used not with constant.

Reply:
There can even occur a static variable! After seeing static function I think it would be a good idea to check also static varibles...

Reply:
I added a test for variables.